### PR TITLE
Implementation of cpp_harness_collect feature

### PR DESCRIPTION
**Pull request by tujh-rf**
Implementation of an issue #89

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -117,12 +117,26 @@ Set it to ``False`` if you have C++ executable files that end with the ``*.py``
...
+cpp_prefix
+^^^^^^^^^^
+
+This option allows the usage of tolls or emulators that are used by invoking them on
+the console wrapping the test binary, like wine or qemu:
+
+.. code-block:: ini
+
+    [pytest]
+    cpp_prefix = wine
```
Comment (nicoddemus) [inline README.rst:129]:
```suggestion
    cpp_prefix = wine -exec
    
The prefix will be used both for listing tests and running them.    
```

I want to make it clear that we support passing options as well (do not know if wine supports some options we can use? used `-exec` just as an example).

File: src/pytest_cpp/boost.py [hunk 2]
```diff
@@ -18,10 +18,19 @@ class BoostTestFacade(AbstractFacade):
...
+        cls,
+        executable: str,
+        prefix: Sequence[str] = (),
+    ) -> bool:
         try:
+            args = list(prefix)
+            args.extend([
+                executable,
+                "--help",
+            ])
```
Comment (nicoddemus) [inline src/pytest_cpp/boost.py:31]:
```suggestion
            args = [*prefix, executable, "--help"]
```

File: src/pytest_cpp/boost.py [hunk 3]
```diff
@@ -52,12 +67,14 @@ def read_file(name: str) -> str:
         with tempfile.TemporaryDirectory(prefix="pytest-cpp") as temp_dir:
             log_xml = os.path.join(temp_dir, "log.xml")
             report_xml = os.path.join(temp_dir, "report.xml")
-            args = list(harness) + [
+            args = list(prefix)
```
Comment (nicoddemus) [inline src/pytest_cpp/boost.py:70]:
```suggestion
            args = [*prefix, *harness, executable, "--output_format=XML",  f"--log_sink={log_xml}", f"--report_sink={report_xml}"]
```

File: src/pytest_cpp/catch2.py [hunk 4]
```diff
@@ -19,10 +19,19 @@ class Catch2Facade(AbstractFacade):
...
 
     @classmethod
-    def is_test_suite(cls, executable: str) -> bool:
+    def is_test_suite(
+        cls,
+        executable: str,
+        prefix: Sequence[str] = (),
+    ) -> bool:
         try:
+            args = list(prefix)
```
Comment (nicoddemus) [inline src/pytest_cpp/catch2.py:28]: Same here using the *syntax to expand the lists. 😁

File: src/pytest_cpp/catch2.py [hunk 5]
```diff
@@ -59,17 +77,20 @@ def run_test(
         executable: str,
         test_id: str = "",
         test_args: Sequence[str] = (),
+        prefix: Sequence[str] = (),
         harness: Sequence[str] = (),
     ) -> tuple[Sequence[Catch2Failure] | None, str]:
         with tempfile.TemporaryDirectory(prefix="pytest-cpp") as tmp_dir:
             xml_filename = os.path.join(tmp_dir, "cpp-report.xml")
-            args = list(harness) + [
+            args = list(prefix)
```
Comment (nicoddemus) [inline src/pytest_cpp/catch2.py:85]: Ditto

File: src/pytest_cpp/google.py [hunk 6]
```diff
@@ -19,10 +19,19 @@ class GoogleTestFacade(AbstractFacade):
...
 
     @classmethod
-    def is_test_suite(cls, executable: str) -> bool:
+    def is_test_suite(
+        cls,
+        executable: str,
+        prefix: Sequence[str] = (),
+    ) -> bool:
         try:
+            args = list(prefix)
```
Comment (nicoddemus) [inline src/pytest_cpp/google.py:28]: Ditto

File: src/pytest_cpp/google.py [hunk 7]
```diff
@@ -41,8 +54,13 @@ def list_tests(self, executable: str) -> list[str]:
           ReturnsTrueForPrimes
           CanGetNextPrime
         """
+        args = list(prefix)
```
Comment (nicoddemus) [inline src/pytest_cpp/google.py:57]: Ditto

File: src/pytest_cpp/google.py [hunk 8]
```diff
@@ -69,15 +87,18 @@ def run_test(
         executable: str,
         test_id: str,
         test_args: Sequence[str] = (),
+        prefix: Sequence[str] = (),
         harness: Sequence[str] = (),
     ) -> tuple[list[GoogleTestFailure] | None, str]:
         with tempfile.TemporaryDirectory(prefix="pytest-cpp") as tmp_dir:
             xml_filename = os.path.join(tmp_dir, "cpp-report.xml")
-            args = list(harness) + [
+            args = list(prefix)
```
Comment (nicoddemus) [inline src/pytest_cpp/google.py:95]: Ditto

File: src/pytest_cpp/boost.py [hunk 9]
```diff
@@ -30,7 +39,12 @@ def is_test_suite(cls, executable: str) -> bool:
         else:
             return "--output_format" in output and "log_format" in output
 
-    def list_tests(self, executable: str) -> list[str]:
+    def list_tests(
+        self,
+        executable: str,
+        prefix: Sequence[str] = (),
```
Comment (nicoddemus) [inline src/pytest_cpp/boost.py:45]: Not sure if it is better to remove the default value for `prefix`... we probably want a hard exception in case we do not pass it explicitly.

**Review by nicoddemus — CHANGES_REQUESTED** [review]
Looks great for a first interaction!

Other than the minor comments, can you think of a test we can have to avoid regressions?

One idea is that we do not really need a full integration test, we might just extract the logic of building the command line into a function, for example today we build the command line like this:



```python
    args = list(prefix)
    args.extend([
        executable,
        "--help",
    ])
```

We could use instead a function with specific parameters that would ensure all required pieces of the pytest-cpp command line are provided:

```python
def make_list_tests_cmdline(prefix: Sequence[str], executable: str, *args: str) -> Sequence[str]:
    ...

def make_run_test_cmdline(prefix: Sequence[str], harness: Sequence[str], executable: str, *args: str) -> Sequence[str]:
    ...    
```

We can then make two quick unittests.

Using those functions cements all the pieces needed to build a command-line, so there's no chance of accidently missing a piece when adding a new command-line. What do you think?

File: README.rst (hunk 1 above)
Comment (tujh-rf) [reply README.rst]: I made example with qemu because wine usually doesn't require arguments

File: src/pytest_cpp/boost.py (hunk 3 above)
Comment (tujh-rf) [reply src/pytest_cpp/boost.py]: solved with additional function `make_cmdline`

File: src/pytest_cpp/boost.py (hunk 9 above)
Comment (tujh-rf) [reply src/pytest_cpp/boost.py]: I do believe exception is not needed here because prefix cpp_harness_collect might be ommited

Comment (nicoddemus) [reply src/pytest_cpp/boost.py]: Not sure I follow... I just meant to remove the default value, so if somebody forgets to pass it Python itself will raise an exception. This is used internally only, so it is best to ensure we are not forgetting to pass `prefix` at some point (and having a default value would let that error slip by).

File: src/pytest_cpp/boost.py [hunk 10]
```diff
@@ -50,15 +61,11 @@ def read_file(name: str) -> str:
...
         with tempfile.TemporaryDirectory(prefix="pytest-cpp") as temp_dir:
-            log_xml = os.path.join(temp_dir, "log.xml")
-            report_xml = os.path.join(temp_dir, "report.xml")
-            args = list(harness) + [
-                executable,
-                "--output_format=XML",
-                "--log_sink=%s" % log_xml,
-                "--report_sink=%s" % report_xml,
-            ]
+            log_xml = os.path.join(os.path.relpath(tmp_dir), "log.xml")
```
Comment (nicoddemus) [inline src/pytest_cpp/boost.py:64]: Are you making the path relative to `tmp_dir`, and I think you intended for the process to execute in the `temp_dir` (seems there is a typo here), but you never pass `cwd` to `subprocess.Popen`, so this will not work.

Also changing the cwd can be considered a breaking change, now that I think more about it...

File: src/pytest_cpp/google.py [hunk 11]
```diff
@@ -72,12 +83,8 @@ def run_test(
         harness: Sequence[str] = (),
     ) -> tuple[list[GoogleTestFailure] | None, str]:
         with tempfile.TemporaryDirectory(prefix="pytest-cpp") as tmp_dir:
-            xml_filename = os.path.join(tmp_dir, "cpp-report.xml")
-            args = list(harness) + [
-                executable,
-                "--gtest_filter=" + test_id,
-                "--gtest_output=xml:%s" % xml_filename,
-            ]
+            xml_filename = os.path.join(os.path.relpath(tmp_dir), "cpp-report.xml")
```
Comment (nicoddemus) [inline src/pytest_cpp/google.py:86]: Same problem here regarding `relpath` and cwd.

File: src/pytest_cpp/helpers.py [hunk 12]
```diff
@@ -0,0 +1,8 @@
+
+from typing import Sequence
+
+
+def make_cmdline(harness: Sequence[str], executable: str, arg: Sequence[str] = ()) ->Sequence[str]:
+    if not executable:
```
Comment (nicoddemus) [inline src/pytest_cpp/helpers.py:6]: Hmm why is this needed?

File: src/pytest_cpp/plugin.py [hunk 13]
```diff
@@ -102,7 +106,13 @@ def pytest_addoption(parser: pytest.Parser) -> None:
...
         type="args",
         default=(),
-        help="command that wraps the cpp binary",
+        help="command that wraps the cpp binary when tests should run",
+    )
+    parser.addini(
+        "cpp_harness_collect",
+        type="args",
+        default=(),
+        help="command that wraps the cpp binary when a list of tests is collecting",
```
Comment (nicoddemus) [inline src/pytest_cpp/plugin.py:115]:
```suggestion
        help="command that wraps the cpp binary when collecting tests",
```

File: src/pytest_cpp/plugin.py [hunk 14]
```diff
@@ -102,7 +106,13 @@ def pytest_addoption(parser: pytest.Parser) -> None:
         "cpp_harness",
         type="args",
         default=(),
-        help="command that wraps the cpp binary",
+        help="command that wraps the cpp binary when tests should run",
```
Comment (nicoddemus) [inline src/pytest_cpp/plugin.py:109]:
```suggestion
        help="command that wraps the cpp binary when running tests",
```

File: tests/test_pytest_cpp.py [hunk 15]
```diff
@@ -91,6 +92,29 @@ def test_success(facade, name, test_id, exes):
     assert facade.run_test(exes.get(name), test_id)[0] is None
 
 
+def test_cmdline_builder_happy_flow():
+    arg_string = make_cmdline( ["wine"], "gtest", ["--help"] )
+    assert arg_string
+    assert arg_string[0] == 'wine', "First element MUST be a `harness` value"
```
Comment (nicoddemus) [inline tests/test_pytest_cpp.py:98]: Better we check the whole list:

```suggestion
    assert arg_string == ["wine", "gtest", "--help"]
```

Same for the other tests.

File: src/pytest_cpp/boost.py (hunk 10 above)
Comment (tujh-rf) [reply src/pytest_cpp/boost.py:64]: I think I commited wrong branch (it is about a typo), will check it.
`cwd` doesn't work in case of using Wine because, again, it will be a local unix-path (/home/user/project/bla-bla-bla/tmp/pytestcpp-qergqr12) and wine can work only with windows format (c:\Users\project\bla-bla-bla\tmp\pytestcpp-qergqr12) however if I'm using relative path for a temp folder (like ../../../tmp/pytestcpp-qergqr12) wine can handle it properly.

File: src/pytest_cpp/google.py (hunk 11 above)
Comment (tujh-rf) [reply src/pytest_cpp/google.py:86]: see my commect upper

File: src/pytest_cpp/helpers.py (hunk 12 above)
Comment (tujh-rf) [reply src/pytest_cpp/helpers.py:6]: ohh, I'm C++ developer much more than Python :)

Comment (nicoddemus) [reply src/pytest_cpp/helpers.py:6]: Hehehe OK we can drop the if then, I think.

File: src/pytest_cpp/boost.py (hunk 10 above)
Comment (nicoddemus) [reply src/pytest_cpp/boost.py:64]: Ahh OK I misunderstood the patch, `relpath` will make a relative path from tmp_dir to cwd already.

We need to account for this though:

> On Windows, ValueError is raised when path and start are on different drives.

In this case, catching the error and falling back to the full path seems reasonable.

File: src/pytest_cpp/boost.py (hunk 9 above)
Comment (tujh-rf) [reply src/pytest_cpp/boost.py]: I did try it, however this require a lot of changes in exist tests because I do not see any reason to remove a default value for a `prefix` in the observation/collecting functions and keep it for a `harness` in a test run, and to be able to do that I need to reorder a list of arguments for `run_test`.

Comment (nicoddemus) [reply src/pytest_cpp/boost.py]: OK

**Comment by tujh-rf** [comment]
```
>       result.stdout.fnmatch_lines(["*%s*" % line])
E       Failed: nomatch: '*5 failed, 1 passed in *'
E           and: '============================= test session starts =============================='
E           and: 'platform linux -- Python 3.7.15, pytest-7.2.1, pluggy-1.0.0'
E           and: 'rootdir: /tmp/pytest-of-runner/pytest-0/test_samples14'
E           and: 'plugins: xdist-3.1.0, cpp-0.1.dev1+g8f10bf5.d20230123, mock-3.10.0'
E           and: 'collected 6 items'
E           and: ''
E           and: '030-Asn-Require-Check FFFFFF                                             [100%]'
E           and: ''
E           and: '=================================== FAILURES ==================================='
E           and: '_____________________ Assert that something is true (pass) _____________________'
E       remains unmatched: '*5 failed, 1 passed in *'
```
I don't get it, does it means the test expected 5 from 6 tests will fail and now all 6 tests failed?
Looks like I made a regression, let me check.

**Comment by tujh-rf** [comment]
found and fixed, my typo

File: src/pytest_cpp/boost.py [hunk 16]
```diff
@@ -50,15 +61,19 @@ def read_file(name: str) -> str:
...
-            args = list(harness) + [
-                executable,
-                "--output_format=XML",
-                "--log_sink=%s" % log_xml,
-                "--report_sink=%s" % report_xml,
-            ]
+            """
+            On Windows, ValueError is raised when path and start are on different drives.
+            In this case failing back to the absolute path.
+            """
```
Comment (nicoddemus) [inline src/pytest_cpp/boost.py:67]:
```suggestion
            # On Windows, ValueError is raised when path and start are on different drives.
            # In this case failing back to the absolute path.
```

File: src/pytest_cpp/google.py [hunk 17]
```diff
@@ -71,13 +82,16 @@ def run_test(
...
-            args = list(harness) + [
-                executable,
-                "--gtest_filter=" + test_id,
-                "--gtest_output=xml:%s" % xml_filename,
-            ]
+        with tempfile.TemporaryDirectory(prefix="pytest-cpp") as temp_dir:
+            """
+            On Windows, ValueError is raised when path and start are on different drives.
+            In this case failing back to the absolute path.
+            """
```
Comment (nicoddemus) [inline src/pytest_cpp/google.py:89]:
```suggestion
            # On Windows, ValueError is raised when path and start are on different drives.
            # In this case failing back to the absolute path.
```

**Comment by nicoddemus** [comment]
Thanks @tujh-rf!

I will prepare a new release soon. :+1:
